egs/swbd/s5c/local/chain/run_discriminative.sh

```diff
@@ -0,0 +1,191 @@
+# since the lattice generation runs in about real-time, so takes of the order of
+# 1000 hours of CPU time.
+# 
+. cmd.sh
+
+
+stage=0
+train_stage=-10 # can be used to start training in the middle.
+get_egs_stage=-10
+use_gpu=true  # for training
+cleanup=false  # run with --cleanup true --stage 6 to clean up (remove large things like denlats,
+               # alignments and degs).
+
+set -e
+. cmd.sh
+. ./path.sh
+. ./utils/parse_options.sh
+
+use_gpu=true  # for training
+srcdir=exp/nnet3/nnet_ms_a
+train_data_dir=data/train_nodup_sp_hires
+online_ivector_dir=exp/nnet3/ivectors_train_nodup_sp
+degs_dir=
+lats_dir=
+
+## Objective options
+criterion=smbr
+one_silence_class=true
+
+## Egs options
+frames_per_eg=150
+frames_overlap_per_eg=30
+truncate_deriv_weights=10
+
+## Nnet training options
+effective_learning_rate=0.0000125
+max_param_change=1
+num_jobs_nnet=4
+num_epochs=4
+regularization_opts=
+minibatch_size=64
+adjust_priors=true
+modify_learning_rates=true
+last_layer_factor=1.0
+
+## Decode options
+decode_start_epoch=1 # can be used to avoid decoding all epochs, e.g. if we decided to run more.
+
+if $use_gpu; then
+  if ! cuda-compiled; then
+    cat <<EOF && exit 1 
+This script is intended to be used with GPUs but you have not compiled Kaldi with CUDA 
+If you want to use GPUs (and have them), go to src/, and configure and make on a machine
+where "nvcc" is installed.  Otherwise, call this script with --use-gpu false
+EOF
+  fi
+  num_threads=1
+else
+  # Use 4 nnet jobs just like run_4d_gpu.sh so the results should be
+  # almost the same, but this may be a little bit slow.
+  num_threads=16
+fi
+
+if [ ! -f ${srcdir}/final.mdl ]; then
+  echo "$0: expected ${srcdir}/final.mdl to exist; first run run_tdnn.sh or run_lstm.sh"
+  exit 1;
+fi
+
+lang=data/lang
+
+if [ $stage -le 1 ]; then
+  # hardcode no-GPU for alignment, although you could use GPU [you wouldn't
+  # get excellent GPU utilization though.]
+  nj=350 # have a high number of jobs because this could take a while, and we might
+         # have some stragglers.
+  use_gpu=no
+  gpu_opts=
+
+  steps/nnet3/align.sh  --cmd "$decode_cmd $gpu_opts" --use-gpu "$use_gpu" \
+     --online-ivector-dir $online_ivector_dir \
+     --scale-opts "--transition-scale=1.0 --acoustic-scale=1.0 --self-loop-scale=1.0" \
+     --nj $nj $train_data_dir $lang $srcdir ${srcdir}_ali || exit 1;
+fi
+
+if [ -z "$lats_dir" ]; then
+  lats_dir=${srcdir}_denlats
+  if [ $stage -le 2 ]; then
+    nj=50  
+    # this doesn't really affect anything strongly, except the num-jobs for one of
+    # the phases of get_egs_discriminative.sh below.
+    num_threads_denlats=6
+    subsplit=40 # number of jobs that run per job (but 2 run at a time, so total jobs is 80, giving
+    # total slots = 80 * 6 = 480.
+    steps/nnet3/make_denlats.sh --cmd "$decode_cmd --mem 1G --num-threads $num_threads_denlats" \
+      --self-loop-scale 1.0 --acwt 1.0 --extra-left-context 20 \
```

Comment:
I can't find where this thread is on github so replying via the email
interface.

Regarding egs/swbd/s5c/local/chain/run_discriminative.sh,
you need to maintain the distinction between 'train' scripts and 'run'
scripts.  (@vijayaditya, take note of this too- you sometimes blur this
distinction).
If you need a general purpose script that takes options, to run
discriminative training, it should go in steps/.  Right now it's not clear
which camp your script falls into.  Although the script is in chain/, you
have srcdir=exp/nnet3/nnet_ms_a, which doesn't look like a chain directory.
If you need a script in local/ which is general-purpose and will be called
from top-level run*.sh scripts (e.g. because of special properties of the
database), it should not be called 'run*.sh'.

So vimal- you need to decide whether your script is a chain script or a run
script.  If a run script, you should not have the extra-left-context.  If a
train script, you need to rename it and have a run script which calls it.

Dan


On Thu, Mar 10, 2016 at 2:34 AM, Vimal Manohar <notifications@github.com>
wrote:

> In egs/swbd/s5c/local/chain/run_discriminative.sh
> <https://github.com/kaldi-asr/kaldi/pull/574#discussion_r55642693>:
>
> > +     --online-ivector-dir $online_ivector_dir \
> > +     --scale-opts "--transition-scale=1.0 --acoustic-scale=1.0 --self-loop-scale=1.0" \
> > +     --nj $nj $train_data_dir $lang $srcdir ${srcdir}_ali || exit 1;
> > +fi
> > +
> > +if [ -z "$lats_dir" ]; then
> > +  lats_dir=${srcdir}_denlats
> > +  if [ $stage -le 2 ]; then
> > +    nj=50
> > +    # this doesn't really affect anything strongly, except the num-jobs for one of
> > +    # the phases of get_egs_discriminative.sh below.
> > +    num_threads_denlats=6
> > +    subsplit=40 # number of jobs that run per job (but 2 run at a time, so total jobs is 80, giving
> > +    # total slots = 80 * 6 = 480.
> > +    steps/nnet3/make_denlats.sh --cmd "$decode_cmd --mem 1G --num-threads $num_threads_denlats" \
> > +      --self-loop-scale 1.0 --acwt 1.0 --extra-left-context 20 \
>
> I can remove the extra-left-context. Although this script is currently for
> tdnn, it needs to be modified to work for lstm. Does it make sense to take
> an option like is_rnn? Or is it better to create a separate recipe for the
> lstm models, which requires extra-left-context etc. for all the stages?
> On Thu, Mar 10, 2016 at 1:32 AM Daniel Povey <notifications@github.com>
> wrote: In egs/swbd/s5c/local/chain/run_discriminative.sh <#574 (comment)
> <https://github.com/kaldi-asr/kaldi/pull/574#discussion_r55638968>>: > +
> --online-ivector-dir $online_ivector_dir \ > + --scale-opts
> "--transition-scale=1.0 --acoustic-scale=1.0 --self-loop-scale=1.0" \ > +
> --nj $nj $train_data_dir $lang $srcdir ${srcdir}_ali || exit 1; > +fi > + >
> +if [ -z "$lats_dir" ]; then > + lats_dir=${srcdir}_denlats > + if [ $stage
> -le 2 ]; then > + nj=50 > + # this doesn't really affect anything strongly,
> except the num-jobs for one of > + # the phases of
> get_egs_discriminative.sh below. > + num_threads_denlats=6 > + subsplit=40
> # number of jobs that run per job (but 2 run at a time, so total jobs is
> 80, giving > + # total slots = 80 * 6 = 480. > +
> steps/nnet3/make_denlats.sh --cmd "$decode_cmd --mem 1G --num-threads
> $num_threads_denlats" \ > + --self-loop-scale 1.0 --acwt 1.0
> --extra-left-context 20 \ there is no need for extra-left-context in
> TDNN-based setups- it makes no difference for TDNNs. — Reply to this email
> directly or view it on GitHub <
> https://github.com/kaldi-asr/kaldi/pull/574/files#r55638968>.
> -- Vimal Manohar PhD Student Electrical & Computer Engineering Johns
> Hopkins University
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/574/files#r55642693>.
>